[IDE] Track whether a file load was initiated by the File Manager better

iainx opened the pull request:
Instead of using a timer that isn't very accurate, track whether or not a
file load was initiated by the FileManager by passing a boolean to the events

Fixes VSTS #806672 https://devdiv.visualstudio.com/DevDiv/_workitems/edit/806672

iainx commented:
@monojenkins backport release-8.0

mrward reviewed (commented):
Should we make all use of InitiatedByFileManager internal? Do we want/need this to be public?

iainx commented:
We don't, no.

iainx commented:
But there were a few places that we couldn't make it internal due to internal not being visible IIRC

Commit:
[IDE] Send TimeToCode event when the files have finished loading

Instead of waiting for a timer to fire to decide whether it was a File manager
initiated load or not, wait for the file manager file load to complete

Fixes VSTS #806672

iainx commented:
@slluis Updated to take your idea into consideration

Therzok commented on main/src/addins/MacPlatform/MacPlatform.cs:
```diff
@@ -628,21 +637,25 @@ void GlobalSetup ()
-					//OpenFiles may pump the mainloop, but can't do that from an AppleEvent, so use a brief timeout
-					GLib.Timeout.Add (0, delegate {
+					//OpenFiles may pump the mainloop, but can't do that from an AppleEvent
+					MonoDevelop.Core.Runtime.RunInMainThread ( () => {
```
Runtime.RunInMainThread is sync if we're already on the main thread.

Therzok replied:
So this changes behaviour.

Therzok commented on main/src/addins/MacPlatform/MacPlatform.cs:
```diff
@@ -628,21 +637,25 @@ void GlobalSetup ()
-						return false;
+							doc => new FileOpenInformation (doc.Key, null, doc.Value, 1, OpenDocumentOptions.DefaultInternal)),
+							null
+						).ContinueWith ((result) => {
```
I keep forgetting the basic TPL APIs, does this continue on the same scheduler (main thread) or thread pool?

Therzok commented on main/src/addins/MacPlatform/MacPlatform.cs:
```diff
@@ -628,21 +637,25 @@ void GlobalSetup ()
 				ApplicationEvents.OpenUrls += delegate (object sender, ApplicationUrlEventArgs e) {
-					GLib.Timeout.Add (0, delegate {
-						IdeApp.ReportTimeToCode = true;
+					Core.Runtime.RunInMainThread (() => {
```
Same here, change of semantics.

slluis commented on main/src/addins/MacPlatform/MacPlatform.cs:
```diff
@@ -628,21 +637,27 @@ void GlobalSetup ()
+					GLib.Idle.Add (delegate {
 						Ide.WelcomePage.WelcomePageService.HideWelcomePageOrWindow ();
-						IdeApp.ReportTimeToCode = true;
+						IdeApp.StartTimeToCodeTimer ();
```
This is not right, it should be measuring time since startup, not since the open message was received.

iainx replied:
It's in theory corrected at the end.

But if we using a timer since startup, and don't receive an open event, then the timer will never be stopped, unless we use yet another timer to track whether we got the open event

slluis replied:
Well, a Stopwatch is not a real timer, it just stores a timestamp, it doesn't matter if we don't stop it.

slluis commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:
```diff
@@ -318,20 +323,43 @@ static void KeyBindingFailed (object sender, KeyBindingFailedEventArgs e)
+				throw new Exception ("SendTimeToCode called before initialisation completed");
+			}
+
+			LoggingService.LogInfo ("Processing TTC");
```
All of those LogInfo should be LogDebug

slluis commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:
```diff
@@ -101,8 +102,25 @@ public static class IdeApp
+			ttcMetadata.AddProperties (startupMetadata);
+
+			LoggingService.LogInfo ("TTC starting");
+			ttcTimeout = GLib.Timeout.Add (ttcDuration, () => {
```
BTW, I think ttcDuration of 6s is too much

iainx replied:
6s was what was needed for my machine when I tested it. I had 3 seconds initially but it was timing out.

slluis replied:
6 seconds between Start Window showing up and getting the open document event? that's a lot. For me there isn't even time to render the Start Window.

slluis commented on main/src/addins/MacPlatform/MacPlatform.cs:
```diff
@@ -582,6 +583,17 @@ void PatchGtkTheme ()
+			if (Projects.Services.ProjectService.IsWorkspaceItemFile (filename)) {
+				return TimeToCodeMetadata.DocumentType.Solution;
+			} else if (Projects.Services.ProjectService.IsSolutionItemFile (filename)) {
+				return TimeToCodeMetadata.DocumentType.File;
```
No no, IsWorkspaceItemFile() will return true for any WorkspaceItem subclass (such as a Solution). IsSolutionItemFile() will return true for any SolutionItem subclass (such as a Project). In both cases it should return a TimeToCodeMetadata.DocumentType.Solution.

slluis commented:
@monojenkins backport release-8.0-integration
